semgrep/semgrep/output.py

```diff
@@ -173,3 +187,178 @@ def build_output(
     else:
         # https://github.com/python/mypy/issues/6366
         raise RuntimeError(f"Unhandled output format: {type(output_format).__name__}")
+
+
+class OutputSettings(NamedTuple):
+    output_format: OutputFormat
+    output_destination: Optional[str]
+    quiet: bool
+    error_on_findings: bool
+
+
+@contextlib.contextmanager
+def managed_output(output_settings: OutputSettings) -> Generator:  # type: ignore
```

Comment:
What's the advantage of a context manager here? Can we just initialize `OutputHandler`s where necessary and continue to let `__main__` be the final arbiter of exceptions and exiting? It seems odd to have our output handler doing a `sys.exit` for us.

Reply:
The point of the context manager was to ensure that `.close()` gets called and that no matter what happens, we output all the errors we've collected so far in a consistent way.

I agree - I don't love having the output manager call `sys.exit`. However, I don't agree that main should be the final arbiter of exceptions - the whole point of this is to allow us to be able to collect things to output that will always be print in the event of a final exception, which I don't think `__main__` is well set up to do.

Reply:
How about we let the `OutputHandler` deal with the output and have `__main__` deal with the exit codes? I.e. call `.close()` in `finally`, but re-raise the exception `ex` in the `except` so `__main__` can grab the code and exit. That way `OutputHandler` doesn't need to worry about `exit_code`, but can still do all the printing.

I believe this should give us the behavior we want, e.g.

```
In [32]: try: 
    ...:     raise Exception('EXCEPTION') 
    ...: except Exception as e: 
    ...:     print('EXCEPT') 
    ...:     raise 
    ...: finally: 
    ...:     print('FINALLY') 
    ...:                                                                                                
EXCEPT
FINALLY
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
<ipython-input-32-04917ed8d528> in <module>
      1 try:
----> 2     raise Exception('EXCEPTION')
      3 except Exception as e:
      4     print('EXCEPT')
      5     raise

Exception: EXCEPTION
```

Reply:
:+1: I like this